themis/spec/symmetric-keys.md

```diff
@@ -0,0 +1,99 @@
+---
+weight: 80
+title:  Symmetric keys
+---
+
+# Symmetric keys
+
+Symmetric keys are used by [**Secure Cell**](../secure-cell/) cryptosystem.
+
+As of Themis 0.13 released in 2020,
+the following symmetric encryption algorithms are supported
+(in a multitude of modes):
+
+  - AES-128 (_deprecated_)
+  - AES-192
+  - AES-256 **(current default)**
+
+Each of the AES flavors has its own specific requirements for the key size.
+
+Secure Cell uses a _key derivation function_ (KDF)
+to transform user-provided key material into a symmetric key of suitable size.
+Therefore, keys used with Secure Cell do not have a particular format:
+they are just arbitrary non-empty byte strings.
+
+## Generating symmetric keys
+
+The recommended way to generate a symmetric key is to use a cryptographically secure pseudorandom number generator
+to obtain a sufficient number of random bytes, and that's your new symmetric key.
+
+{{< hint danger >}}
+**Warning:**
+It is crucial to use a *cryptographically secure* RNG so that the key is unpredictable.
+Do not use general-purpose random number generators – typically found in “math” libraries –
+unless they are explicitly designed for cryptographic purposes.
+{{< /hint >}}
+
+As of 2020,
+[NIST recommends](https://www.keylength.com/en/4/)
+symmetric keys to be at least 128 bits long for them to be future-proof.
+Themis currently generates 256-bit keys by default (i.e., 32 bytes).
+
+See also our [key management guide](/themis/crypto-theory/key-management/)
+for advice on securing the key after you have generated it.
+
+### Example
+
+Here is a snippet in Go, illustrating how to generate a good symmetric key:
+
+```go
+package main
+
+import (
+	"crypto/rand" // Use a cryptograpic RNG.
+	"fmt"
+)
+
+func main() {
+	// Generate keys of sufficient length.
+	const keyBytes = 32
+	key := make([]byte, keyBytes)
+	_, err := rand.Read(key)
+	if err != nil {
+		// Not handling RNG failure is a common critical error.
+		// If errors are ignored, you might return a predictable key,
+		// such as "all zeros" initial value of the byte array.
+		panic("failed to generate a key")
+	}
+	fmt.Printf("✅ symmetric key: %x\n", key)
+}
+```
+
+## Passphrases
+
+Alternatively, Secure Cell can work with passphrases.
+From the cryptographic point of view, a passphrase is essentially a symmetric key with lower entropy per byte.
+Themis encodes textual passphrases in UTF-8 and uses the resulting byte string.
+
+Since passphrases are typically shorter than keys and have less entropy (randomness),
+Secure Cell uses additional KDF rounds to preprocess passphrase data into an encryption key.
```

Comment:
```suggestion
Secure Cell uses additional PBKDF2 rounds to preprocess passphrase data into an encryption key.
```

let's say somewhere that we use PBKDF2?

Reply:
I've explicitly avoided mentioning PBKDF2 here since it's only an overview of keys and passphrases, not Secure Cell operation. I believe, intricacies of the KDF choice belong to Secure Cell description specifically. For one, it's not only PBKDF2 there, ZRTP-style KDF is also thrown into the mix. It will be easier to have KDF description in one place so that when a new KDF is added to Secure Cell then it's not necessary to locate and edit a bunch of other places as well. Passphrase content does not really depend on the specific KDF in use.

Reply:
If not mention PBKD2 explicitly, maybe let's link to a page where we describe usage of PBKDF2?

Reply:
> maybe let's link to a page where we describe usage of PBKDF2?

As soon as it there. My idea was to cross-link Secure Cell spec with key/passphrase spec once it's actually ready.

Reply:
We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs. https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

Reply:
> We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs.
> https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

Okay, I've added links to this section.